src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -300,11 +279,94 @@ private ValueTask<Optional<TElement>> TryDequeueCoreAsync()
                         _waiters = new Queue<TaskCompletionSource<Optional<TElement>>>();
                     }
 
-                    var waiter = new TaskCompletionSource<Optional<TElement>>();
+                    // Continuations run asynchronously to ensure user code does not execute within protected regions.
+                    var waiter = new TaskCompletionSource<Optional<TElement>>(TaskCreationOptions.RunContinuationsAsynchronously);
+                    AttachCancellation(waiter, cancellationToken);
```

Comment:
This essentially means we are adding an allocation to this path when passing a valid `CancellationToken` value. Given the comments about perf concern with boxing are we worried about the perf of adding this extra allocation here?

Reply:
We added allocation(s) here, but we took at least two away from the `WithCancellationAsync` path (one for `ContinueWith` and one for `Unwrap`). I'm not sure exactly what the net change is, but based acceptability of this allocation on the fact that the previous code also always allocated on the same path.